src/ray/core_worker/task_event_buffer.cc

```diff
@@ -0,0 +1,193 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/task_event_buffer.h"
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+TaskEventBufferImpl::TaskEventBufferImpl(std::unique_ptr<gcs::GcsClient> gcs_client)
+    : work_guard_(boost::asio::make_work_guard(io_service_)),
+      periodical_runner_(io_service_),
+      gcs_client_(std::move(gcs_client)) {}
+
+Status TaskEventBufferImpl::Start(bool auto_flush) {
+  auto report_interval_ms = RayConfig::instance().task_events_report_interval_ms();
+  RAY_CHECK(report_interval_ms > 0)
+      << "RAY_task_events_report_interval_ms should be > 0 to use TaskEventBuffer.";
+
+  buffer_.reserve(RayConfig::instance().task_events_max_num_task_events_in_buffer());
+
+  io_thread_ = std::thread([this]() {
+#ifndef _WIN32
+    // Block SIGINT and SIGTERM so they will be handled by the main thread.
+    sigset_t mask;
+    sigemptyset(&mask);
+    sigaddset(&mask, SIGINT);
+    sigaddset(&mask, SIGTERM);
+    pthread_sigmask(SIG_BLOCK, &mask, NULL);
+#endif
+    SetThreadName("task_event_buffer.io");
+    io_service_.run();
+    RAY_LOG(INFO) << "Task event buffer io service stopped.";
+  });
+
+  // Reporting to GCS, set up gcs client and and events flushing.
+  auto status = gcs_client_->Connect(io_service_);
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << "Failed to connect to GCS, TaskEventBuffer will stop now. [status="
+                   << status.ToString() << "].";
+
+    // Early clean up resources.
+    gcs_client_.reset();
+    Stop();
+    return status;
+  }
+
+  if (!auto_flush) {
+    return Status::OK();
+  }
+
+  RAY_LOG(INFO) << "Reporting task events to GCS every " << report_interval_ms << "ms.";
+  periodical_runner_.RunFnPeriodically([this] { FlushEvents(/* forced */ false); },
+                                       report_interval_ms,
+                                       "CoreWorker.deadline_timer.flush_task_events");
+  return Status::OK();
+}
+
+void TaskEventBufferImpl::Stop() {
+  RAY_LOG(INFO) << "Shutting down TaskEventBuffer.";
+
+  // Shutting down the io service to exit the io_thread. This should prevent
+  // any other callbacks to be run on the io thread.
+  io_service_.stop();
+  if (io_thread_.joinable()) {
+    RAY_LOG(DEBUG) << "Joining io thread from TaskEventBuffer";
+    io_thread_.join();
+  }
+
+  {
+    absl::MutexLock lock(&mutex_);
+    // It's now safe to disconnect the GCS client since it will not be used by any
+    // callbacks.
+    if (gcs_client_) {
+      // Stop GCS client
+      gcs_client_->Disconnect();
+    }
+
+    gcs_client_.reset();
```

Comment:
Why don't we just not reset? This means we should always check the pointer when gcs_client is used.

Reply:
Is there any drawback for keeping the rpc client around? Maybe not - yeah, I guess we could keep it not reset.